From a65303a8f3a4f813c6f4154e5f5ba72e0280bcb2 Mon Sep 17 00:00:00 2001
From: Ikey Doherty <ikey@solus-project.com>
Date: Sat, 5 Nov 2016 13:21:32 +0000
Subject: [PATCH] applets/keyboard-layout: Fallback to locale based language
 detection

In the display string, for the zh_CN locale, we did not manage to retrieve
a language name from libgnome-desktop. This resulted in rendering an odd
label for libpinyin ibus engine, i.e. "(null) (libpinyin)".

When we fail to get a language based on the language code, we should fall
back to locale detection, which should render in the appropriate locale
too.

Signed-off-by: Ikey Doherty <ikey@solus-project.com>
---
 src/panel/applets/keyboard-layout/KeyboardLayoutApplet.vala | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/panel/applets/keyboard-layout/KeyboardLayoutApplet.vala b/src/panel/applets/keyboard-layout/KeyboardLayoutApplet.vala
index 36e71fa..982f5fa 100644
--- a/panel/applets/keyboard-layout/KeyboardLayoutApplet.vala
+++ b/panel/applets/keyboard-layout/KeyboardLayoutApplet.vala
@@ -166,7 +166,10 @@ class InputSource
         }
 
         /* Get useful display string */
-        string language = Gnome.get_language_from_code(engine.language, null);
+        string? language = Gnome.get_language_from_code(engine.language, null);
+        if (language == null) {
+            language = Gnome.get_language_from_locale(engine.language, null);
+        }
         this.description = "%s (%s)".printf(language, engine.name);
 
         string? e_variant = engine.layout_variant;
-- 
2.10.2

